pkg/istiod/istio_ca.go

```diff
@@ -0,0 +1,371 @@
+
+	// This value can also be extracted from the mounted token
+	trustedIssuer = env.RegisterStringVar("TOKEN_ISSUER", "",
+		"OIDC token issuer. If set, will be used to check the tokens.")
+
+	audience = env.RegisterStringVar("AUDIENCE", "istio-ca",
+		"Expected audience in the tokens. For backward compat, default is istio-ca.")
+)
+
+const (
+	bearerTokenPrefix = "Bearer "
+	httpAuthHeader    = "authorization"
+	identityTemplate  = "spiffe://%s/ns/%s/sa/%s"
+)
+
+type CAOptions struct {
+	// domain to use in SPIFFE identity URLs
+	TrustDomain string
+}
+
+// RunCA will start the cert signing GRPC service on an existing server.
+func RunCA(grpc *grpc.Server, cs kubernetes.Interface, opts *CAOptions) {
+	ca := createCA(cs.CoreV1(), opts)
+
+	iss := trustedIssuer.Get()
+	aud := audience.Get()
+
+	ch := make(chan struct{})
+	if token, err := ioutil.ReadFile(JWTPath); err != nil {
+		// for debug we may want to override this by setting trustedIssuer explicitly
+		if iss == "" {
+			log.Warna("istiod running without access to K8S tokens. Disable the CA functionality",
+				JWTPath)
+			return
+		}
+	} else {
+
+		tok, err := detectAuthEnv(string(token))
+		if err != nil {
+			log.Warna("Starting with invalid K8S JWT token", err, string(token))
+		} else {
+			if iss == "" {
+				iss = tok.Iss
+			}
+			if len(tok.Aud) > 0 {
+				aud = tok.Aud[0]
+			}
+		}
+	}
+
+	// The CA API uses cert with the max workload cert TTL.
+	// 'hostlist' must be non-empty - but is not used since a grpc server is passed.
+	caServer, startErr := caserver.NewWithGRPC(grpc, ca, maxWorkloadCertTTL.Get(),
+		false, []string{"istiod.istio-system"}, 0, spiffe.GetTrustDomain(),
+		true)
+	if startErr != nil {
+		log.Fatalf("failed to create istio ca server: %v", startErr)
+	}
+
+	// TODO: if not set, parse Istiod's own token (if present) and get the issuer. The same issuer is used
+	// for all tokens - no need to configure twice. The token may also include cluster info to auto-configure
+	// networking properties.
+	if iss != "" && // issuer set explicitly or extracted from our own JWT
+		(k8sInCluster.Get() != "" || trustedIssuer.Get() != "") { // either set explicitly, or not running in cluster.
+		// Add a custom authenticator using standard JWT validation, if not running in K8S
+		// When running inside K8S - we can use the built-in validator, which also check pod removal (invalidation).
+		oidcAuth, err := newJwtAuthenticator(iss, opts.TrustDomain, aud)
+		if err == nil {
+			caServer.Authenticators = append(caServer.Authenticators, oidcAuth)
+			log.Infoa("Using out-of-cluster JWT authentication")
+		} else {
+			log.Infoa("K8S token doesn't support OIDC, using only in-cluster auth")
+		}
+	}
+
+	if serverErr := caServer.Run(); serverErr != nil {
+		// stop the registry-related controllers
+		ch <- struct{}{}
+
+		log.Warnf("Failed to start GRPC server with error: %v", serverErr)
+	}
+	log.Info("Istiod CA has started")
+}
+
+type jwtAuthenticator struct {
```

Comment:
I've tried to reuse existing code to minimize disruption - but for newly written code I would rather use the top level pkg, in particular since this will also likely be used to authenticate the request to the XDS and MCP servers ( i.e. the pilot and galley portions of the code).